scripts/dump_mysql_serlo_database.ts

```diff
@@ -0,0 +1,200 @@
+  }
+}
+
+abstract class LineTransform extends StringTransform {
+  private unfinishedLine: string = ''
+
+  abstract transformLine(line: string): void
+
+  transformString(newText: string): void {
+    const text = this.unfinishedLine + newText
+    let start: number = 0
+    let end: number = 0
+
+    while (true) {
+      start = end
+      end = text.indexOf('\n', start) + 1
+
+      if (start < end) {
+        this.transformLine(text.substring(start, end))
+      } else {
+        this.unfinishedLine = text.substring(start)
+        break
+      }
+    }
+  }
+
+  _flush(callback: Function) {
+    if (this.unfinishedLine) {
+      this.transformLine(this.unfinishedLine)
+    }
+
+    callback()
+  }
+}
+
+class ConcatenateInsertCommands extends LineTransform {
+  private lastTable: string = ''
+  private currentCmdLength: number = 0
+  private maxInsertCmdLength: number
+
+  constructor(encoding: string, maxInsertCmdLength: number) {
+    super(encoding)
+    this.maxInsertCmdLength = maxInsertCmdLength
+  }
+
+  transformLine(line: string): void {
+    const insertTerm = 'INSERT INTO `'
+    const tableEndTerm = '` ('
+    const valuesTerm = ') VALUES ('
+
+    if (line.startsWith(insertTerm)) {
+      assert.ok(
+        line.endsWith('\n'),
+        'This implementation asserts that an INSERT command is never the last line.'
+      )
+
+      const tableStart = insertTerm.length + 1
+      const tableEnd = line.indexOf(tableEndTerm)
+      const table = line.substring(tableStart, tableEnd)
+
+      const valuesStart = line.indexOf(valuesTerm) + valuesTerm.length - 1
+      const valuesEnd = line.length - ';\n'.length
+      const values = line.substring(valuesStart, valuesEnd)
+
+      if (table !== this.lastTable) {
+        this.lastTable = table
+
+        const insertCommandPrefix = line.substring(0, valuesStart - 1)
+        this.pushInsertCommand(insertCommandPrefix + '\n')
+      } else {
+        this.continueInsertCommand()
+      }
+
+      this.pushInsertCommand(values)
+
+      if (this.currentCmdLength > this.maxInsertCmdLength) {
+        this.endInsertCommand()
+      }
+    } else {
+      this.endInsertCommand()
+      this.push(line)
+    }
+  }
+
+  private pushInsertCommand(part: string) {
+    this.push(part)
+    this.currentCmdLength += part.length
+  }
+
+  private continueInsertCommand(): void {
+    this.pushInsertCommand(',\n')
+  }
+
+  private endInsertCommand(): void {
+    if (this.lastTable) {
+      this.pushInsertCommand(';\n')
+      this.lastTable = ''
+      this.currentCmdLength = 0
+    }
+  }
+}
+
+class IgnoreInsecurePasswordWarning extends LineTransform {
+  private readonly uneccessaryError =
+    'mysqldump: [Warning] Using a password on the command line ' +
+    'interface can be insecure.\n'
+
+  transformLine(line: string): void {
+    if (line !== this.uneccessaryError) {
+      this.push(line)
+    }
+  }
+}
+
+const mysqldump = spawn('docker-compose', dockerComposeArgs)
```

Comment:
Something I like to do when coding longer functions / scripts: I usually try to increase the abstraction level when reading the file from top to bottom. This way, you don't necessarily have to read the whole source file but can stop at the appropriate level (resp. jump only down to the lower level abstractions you need to make your changes and can ignore most other stuff). So for example, in this case I'd roughly order the stuff the following way:

```.ts
// All imports (since they have to be top)

// Any global variables I might need. These must be before their first usage. If I need variables only in some lower level abstraction, move them in the function

// I call exec on top-level before its declaration so that it's clear that this is a sc
exec()

// Define exec (async if this returns a promise)
async function exec() {
  const dockerComposeArgs
  const mysqldump = spawn('docker-compose', dockerComposeArgs)

  mysqldump.stdout
    .pipe(new ConcatenateInsertCommands(encoding, maxInsertCmdLength))
    .pipe(fs.createWriteStream(sqlInitFile))

  mysqldump.stderr
    .pipe(new IgnoreInsecurePasswordWarning('utf8'))
    .pipe(process.stderr)

  mysqldump.on('error', error => {
    console.error('ERROR: ' + error)
  })
  mysqldump.on('exit', (code, signal) => {
    process.exit(code !== null ? code : 1)
  })
}

// Now come all the needed classes / functions, might be even part of the function body itself
// Also sorted in the same way. I.e. on the top are the classes that are used by exec, further low any classes they inherit from
// "Level 1": direct usage in async
class ConcatenateInsertCommands extends LineTransform {}
class IgnoreInsecurePasswordWarning extends LineTransform {}
// "Level 2": super classes of Level 1 classes
abstract class LineTransform extends StringTransform {}
// "Level 3" super classes of Level 2 classes
abstract class StringTransform extends Transform {}
```

Depending on how complex the stuff you are doing is, you might want to introduce additional levels, e.g.
```.ts
async function exec() {
  const dump = dumpMysqlDatabase()
  concatInsertCommands(dump)
  ignoreInsecurePasswordWarning(dump)

  function concatInsertCommands(dump) {}
  function ignoreInsecurePasswordWarning(dump) {}
}
```

See also https://github.com/edtr-io/edtr-io/blob/dbd08bad69cdea55c83d4f6fc9d152014e5a0ee8/packages/private/demo/scripts/deploy.ts for another example.

Note: this approach only works for `class` and `function`. Any variable defined with `const` or `let` and therefore any arrow function `const f = () => {}` has to be defined before their first usage. But you can at least move them in a deeper scope resp. abstract them away in functions where it makes sense.

Reply:
@kulla FYI, some suggestion on how to improve code readability